qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,306 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from base64 import urlsafe_b64decode
+from string import ascii_letters, digits
+import datetime
+from random import SystemRandom
+import functools
+from traceback import format_exception
+
+from tornado.web import RequestHandler
+
+from moi import r_client
+from qiita_core.exceptions import (IncorrectPasswordError, IncorrectEmailError,
+                                   UnverifiedEmailError)
+import qiita_db as qdb
+
+
+def login_client(client_id, client_secret=None):
+    """Login client_id, optionally with client_secret
+
+    Parameters
+    ----------
+    client_id : str
+        The client ID to validate
+    client_secret : str, optional
+        The client secret code. If not given, will only validate if client_id
+        exists
+
+    Returns
+    -------
+    bool
+        Whether client_id exists and, if client_secret given, the secret
+        matches the id
+    """
+    with qdb.sql_connection.TRN:
+        sql = """SELECT EXISTS(
+                    SELECT *
+                    FROM qiita.oauth_identifiers
+                    WHERE client_id = %s {0})"""
+        sql_info = [client_id]
+        if client_secret is not None:
+            sql = sql.format("AND client_secret = %s")
+            sql_info.append(client_secret)
+        else:
+            sql = sql.format("AND client_secret IS NULL")
+        qdb.sql_connection.TRN.add(sql, sql_info)
+        return qdb.sql_connection.TRN.execute_fetchlast()
+
+
+def _oauth_error(self, error_msg, error):
+            self.set_status(400)
+            self.write({'error': error,
+                        'error_description': error_msg})
+            self.finish()
+
+
+def authenticate_oauth(f):
+    """Decorate methods to require valid Oauth2 Authorization header[1]
+
+    If a valid header is given, the handoff is done and the page is rendered.
+    If an invalid header is given, a 400 error code is returned and the json
+    error message is automatically sent.
+
+    References
+    ---------
+    [1] The OAuth 2.0 Authorization Framework.
+    http://tools.ietf.org/html/rfc6749
+    """
+    @functools.wraps(f)
+    def wrapper(self, *args, **kwargs):
+        header = self.request.headers.get('Authorization', None)
+        if header is None:
+            _oauth_error(self, 'Oauth2 error: invalid access token',
+                         'invalid_request')
+            return
+        token_info = header.split()
+        if len(token_info) != 2 or token_info[0] != 'Bearer':
```

Comment:
Agree with @antgonza on adding a bit more points of documentation. @squirrelo we are not asking to duplicate the entire RFC, but you can't expect that every single developer will read the RFC in order to understand the code that is written here.

As an example of documentation that I would expect having here can be seen here: https://github.com/idan/oauthlib/blob/master/oauthlib/oauth2/rfc6749/request_validator.py

Also, looking at that - the question that raises is why did you re-implement the RFC if the oauthlib already includes an implementation-independent framework for OAuth2? We shouldn't be reinventing the wheel, specially on topics that we are not experts.